maven-core/src/main/java/org/apache/maven/plugin/internal/ValidatingConfigurationListener.java

```diff
@@ -94,4 +111,53 @@ private void notify( String fieldName, Object value )
         }
     }
 
+    private void warnDeprecated( String fieldName, Object value )
+    {
+        Parameter parameter = mojoDescriptor.getParameterMap().get( fieldName );
+        String deprecated = parameter.getDeprecated();
+        if ( deprecated != null && !deprecated.isEmpty() )
+        {
+            if ( !toString( value ).equals( toString( parameter.getDefaultValue() ) ) )
+            {
+                StringBuilder sb = new StringBuilder( "  Parameter '" );
+                sb.append( fieldName ).append( '\'' );
+                if ( parameter.getExpression() != null )
+                {
+                    String userProperty = parameter.getExpression().replace( "${", "'" ).replace( '}', '\'' );
+                    sb.append( " (User Property " ).append( userProperty ).append( ")" );
+                }
+                sb.append( " is deprecated. " ).append( deprecated );
+
+                logger.warn( MessageUtils.buffer().warning( sb.toString() ).toString() );
+            }
+        }
+    }
+
+    /**
+     * Creates a human-friendly string representation of the specified object.
+     *
+     * @param obj The object to create a string representation for, may be <code>null</code>.
+     * @return The string representation, never <code>null</code>.
+     */
+    private String toString( Object obj )
```

Comment:
can be static

Reply:
I agree one class is simpler.

Reply:
DebugConfigurationListener can not be removed just yet (it is used in m-assembly-p)

Reply:
https://github.com/apache/maven-assembly-plugin/blob/master/src/main/java/org/apache/maven/plugins/assembly/archive/DefaultAssemblyArchiver.java#L439

https://github.com/apache/maven-mvnd/blob/055e50feb52ff3155b7fcf254e99901e1928c7fd/daemon/src/main/java/org/mvndaemon/mvnd/plugin/CliMavenPluginManager.java#L555

So maybe better will be use `DebugConfigurationListener` instead of duplicate code?

Reply:
I have no problem delegating to `DebugConfigurationListener` instead of duplicating code in `ValidatingConfigurationListener`, but in that case I suggest also to "un-deprecate" `DebugConfigurationListener` because it is just confusing if that class should continue to be used or not.

Reply:
Ok, today we can delegate to `DebugConfigurationListener` ... it will be another issue to eventually remove it.